src/makeConfig.ts

```diff
@@ -0,0 +1,69 @@
+import { Options, FetchContextTypes, OptionsMaybeURL, NoUrlOptions } from './types'
+import { isString, isObject, pullOutRequestInit } from "./utils"
+
+export const DefaultOptions: NoUrlOptions = {
+  onMount: false
+}
+
+function makeConfig(context: FetchContextTypes, urlOrOptions?: string | OptionsMaybeURL, optionsNoURLs?: NoUrlOptions): Options {
+  let options: Partial<Options> = {}
+  let requestInit: RequestInit = {}
+
+  if (isObject(urlOrOptions) && isObject(optionsNoURLs)) {
+    // TODO: I could not get the utils/invariant to work with the condition isObject(urlOrOptions) && isObject(optionsNoURLs)
+    throw new Error('You cannot have a 2nd parameter of useFetch when your first argument is a object config.')
+  }
+
+  const getUrl = (): string => {
+    if (isString(urlOrOptions)) {
+      return urlOrOptions
+    }
+
+    if (isObject(urlOrOptions) && !!urlOrOptions.url) {
+      return urlOrOptions.url;
+    }
+
+    if (isObject(urlOrOptions) && !!urlOrOptions.url) {
+      return urlOrOptions.url;
+    }
+
+    if (!!context.url) {
+      return context.url;
+    }
+
+    // we need to throw rather than using invariant so getUrl does not return a value and casts to never
+    throw new Error('You have to either set a URL in your options config or set a global URL in your <Provider url="https://url.com"></Provider>')
+  }
+
+  const getOnMount = (): boolean => {
+    if (isObject(urlOrOptions)) {
+      return !!urlOrOptions.onMount
+    }
+
+    if (isObject(optionsNoURLs)) {
+      return !!optionsNoURLs.onMount
+    }
+
+    return false;
+  }
+
+  const requestInitOptions = isObject(urlOrOptions) ? urlOrOptions : isObject(optionsNoURLs) ? optionsNoURLs : {}
+  requestInit = pullOutRequestInit(requestInitOptions)
+
+  return {
+    url: getUrl(),
+    onMount: getOnMount(),
+    headers: {
+      // default content types http://bit.ly/2N2ovOZ
+      // Accept: 'application/json', 
+      'Content-Type': 'application/json',
+      ...(context.options || {}).headers,
+      ...options.headers
+    },
+    ...options,
+    ...requestInit,
+  } as Options
+}
+
+export { makeConfig }
+export default makeConfig
```

Comment:
I like this. It makes the code really clean

Reply:
thank you.  isolated and more testable.

Reply:
We will need to change this a little with my changes for my next PR. It's not a huge deal, but basically what we need to change is when submitting FormData, we need to remove the `Content-Type: 'application/json'` header. I think that's it for the most part

Reply:
should we not get one merged into master first and then apply master to the other one?

Reply:
Yes we should. merge conflicts are gonna suck 😭